### DISCUSSION TITLE
flash version of pushImage with color map and transparency

### DISCUSSION THREAD

**Discussion by ygator** [Bodmer/TFT_eSPI#883] [New processor/display support or capability request] [answered]
I have created a program that converts an animated gif into PROGMEM data for use with pushImage w/ transparency flash version.  I just saw the colormap version and was thinking of changing my program to use it as it would be way less data.  However, I see there is no flash version of pushImage with color map.  Is there a reason this was left out?

Thanks.

**Comment by Bodmer** [comment, marked as answer]
There are[ suitable functions here](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.h#L482-L486). As follows:
```
           // These are used by Sprite class pushSprite() member function for 1, 4 and 8 bits per pixel (bpp) colours
           // They are not intended to be used with user sketches (but could be)
           // Set bpp8 true for 8bpp sprites, false otherwise. The cmap pointer must be specified for 4bpp
  void     pushImage(int32_t x, int32_t y, int32_t w, int32_t h, uint8_t  *data, bool bpp8 = true, uint16_t *cmap = nullptr);
  void     pushImage(int32_t x, int32_t y, int32_t w, int32_t h, uint8_t  *data, uint8_t  transparent, bool bpp8 = true, uint16_t *cmap = nullptr);
```
You would have to cast the pointer to a uint8_t* type to use them.  Only 4 bits per pixel images have a colour map (palette). For a 4bpp colour set bpp8 to false and include a colour map.

**Reply by ygator** [reply]
> Only 4 bits per pixel images have a colour map (palette). For a 4bpp colour set bpp8 to false and include a colour map.

Thanks, I missed that.  I was thinking I could pass the gif bitmap which is 8bpp as the data and translate the global and local color tables from rgb888 to rgb565 and pass that as the cmap.

I'll take a look at the other items you suggested, but I may look at creating a pushImage working as I mentioned above.

Thanks again.

**Comment by Bodmer** [comment]
P.S, there is a [GIF decoder library here](https://github.com/bitbank2/AnimatedGIF) that works quite well and I added the [TFT_eSPI example](https://github.com/bitbank2/AnimatedGIF/tree/master/examples/TFT_eSPI_memory) recently that may be of interest!
